### PR TITLE
OHRM5X-1449: Develop performance - define workflows for review

### PR DESCRIPTION
**Pull request by devishke-orange**
PR contains:
- OHRM5X-1449: [BE] Define workflow state machine

image
image

### CODE REVIEW

File: installer/Migration/V5_1_0/permission/api.yaml [hunk 1]
```diff
@@ -128,32 +128,33 @@ apiv2_performance_performance_tracker_logs:
...
     create: false
     update: false
     delete: false
   permissions:
-    - { role: Admin, permission: { read: true,  create: false, update: false, delete: false } }
-    - { role: Supervisor, permission: { read: true,  create: false, update: false, delete: false } }
-    - { role: ESS, permission: { read: true,  create: false, update: false, delete: false } }
+    - { role: Admin, permission: { read: true, create: false, update: false, delete: false, self: false } }
+    - { role: Supervisor, permission: { read: true, create: false, update: false, delete: false, self: false } }
+    - { role: ESS, permission: { read: true, create: false, update: false, delete: false, self: false } }
```
Comment (RajithaKumara) [inline installer/Migration/V5_1_0/permission/api.yaml:157]:
```suggestion
    - { role: Admin, permission: { read: true, create: false, update: false, delete: false } }
    - { role: Supervisor, permission: { read: true, create: false, update: false, delete: false } }
    - { role: ESS, permission: { read: true, create: false, update: false, delete: false } }
```
@devishke-orange @SharukaNP @SandeepaOHRM 
will remove self from all APIs, default is false. Also no usage of self in API. but don't remove from other data groups. remove only for API data groups with prefix `apiv2_`

File: src/plugins/orangehrmPerformancePlugin/Api/PerformanceReviewAllowedActionsAPI.php [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+     * @param PerformanceReview $performanceReview
+     * @return int
+     */
+    private function getPerformanceReviewStatus(PerformanceReview $performanceReview): int
+    {
+        if ($this->getAuthUser()->getEmpNumber() === $performanceReview->getEmployee()->getEmpNumber()) {
+            $selfReviewer = $this->getPerformanceReviewService()
+                ->getPerformanceReviewDao()
+                ->getPerformanceSelfReviewer($performanceReview);
+            return $selfReviewer->getStatus() + 1;
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPerformancePlugin/Api/PerformanceReviewAllowedActionsAPI.php:128]: not clear

File: src/plugins/orangehrmPerformancePlugin/test/Api/PerformanceReviewAllowedActionsAPITest.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+namespace OrangeHRM\Tests\Performance\Api;
+
+use OrangeHRM\Framework\Services;
+use OrangeHRM\Performance\Api\PerformanceReviewAllowedActionsAPI;
+use OrangeHRM\Tests\Util\EndpointIntegrationTestCase;
+use OrangeHRM\Tests\Util\Integration\TestCaseParams;
+
+class PerformanceReviewAllowedActionsAPITest extends EndpointIntegrationTestCase
```
Comment (RajithaKumara) [inline src/plugins/orangehrmPerformancePlugin/test/Api/PerformanceReviewAllowedActionsAPITest.php:27]: add group

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmPerformancePlugin/Api/PerformanceReviewAllowedActionsAPI.php (hunk 2 above)
Comment (devishke-orange) [reply src/plugins/orangehrmPerformancePlugin/Api/PerformanceReviewAllowedActionsAPI.php:128]: self status is 1 (activated), 2 (in progress), 3 (completed). Overall status is 1 (inactive), 2 (activated), 3 (in progress), 4 (completed). I have mapped label names with overall status in `ACTIONABLE_STATES_MAP`. that's why I add 1 here to get the proper status.

I'll add a comment here to make it more clear

File: src/plugins/orangehrmPerformancePlugin/test/Api/PerformanceReviewAllowedActionsAPITest.php (hunk 3 above)
Comment (devishke-orange) [reply src/plugins/orangehrmPerformancePlugin/test/Api/PerformanceReviewAllowedActionsAPITest.php:31]: group added

File: src/plugins/orangehrmPerformancePlugin/Dao/PerformanceReviewDao.php [hunk 4]
```diff
@@ -423,19 +426,12 @@ public function getReviewIdList(): array
      */
     public function getSelfReviewIds(int $employeeNumber): array
     {
-        $q = $this->createQueryBuilder(Reviewer::class, 'reviewer');
```
Comment (SharukaNP) [inline src/plugins/orangehrmPerformancePlugin/Dao/PerformanceReviewDao.php:414]: Good

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**